File: apps/sintraprime/src/governance/trustAuthorityRouter.ts
/**
 * Governed Trust Authority Router
 *
 * Enforces the Howard Trust Authority Stack at the SintraPrime control plane.
 * The router is intentionally fail-closed for legal-effect conclusions and
 * external execution. It does not pretend that secondary educational material
 * is controlling law and it does not fabricate a current-law verification.
 */

import type { PlanStep, TaskRequest } from '../types/index.js';

export type TrustAuthorityStage =
  | 'trust-instrument-authority'
  | 'weisss-trustee-handbook'
  | 'current-law-verifier';

export interface TrustAuthorityVerification {
  status: 'VERIFIED_CURRENT' | 'NOT_YET_VERIFIED' | 'CONFLICT_FOUND';
  jurisdiction?: string;
  authorities?: string[];
  verifiedAt?: string;
  verifier?: string;
}

export interface TrustAuthorityRoute {
  isTrustRelated: boolean;
  routeId: 'HOWARD-TRUST-AUTHORITY';
  authorityOrder: TrustAuthorityStage[];
  legalEffectRequested: boolean;
  externalExecutionRequested: boolean;
  currentLawVerification: TrustAuthorityVerification;
  principalApproval: boolean;
  executionAllowed: boolean;
  blockingReasons: string[];
}

const TRUST_PATTERNS = [
  /\bisiah\s+tarik\s+howard\s+trust\b/i,
  /\btrust instrument\b/i,
  /\bcertification of trust\b/i,
  /\bdeclaration of trust\b/i,
  /\btrustee\b/i,
  /\bco-?trustee\b/i,
  /\bbeneficiar(?:y|ies)\b/i,
  /\bsettlor\b/i,
  /\btrust corpus\b/i,
  /\bfiduciary dut(?:y|ies)\b/i,
  /\btrust administration\b/i,
  /\btrust banking\b/i,
  /\btrust resolution\b/i,
  /\btrust amendment\b/i,
];

const LEGAL_EFFECT_PATTERNS = [
  /\blegal effect\b/i,
  /\blegally binding\b/i,
  /\benforce(?:able|ment)?\b/i,
  /\bperfect(?:ion|ed)?\b/i,
  /\blien\b/i,
  /\bjurisdiction\b/i,
  /\btax(?:able|ation| filing| status)?\b/i,
  /\bcourt\b/i,
  /\bcreditor\b/i,
  /\bbank obligation\b/i,
  /\bborrow(?:ing)?\b/i,
  /\bencumber\b/i,
  /\bdistribut(?:e|ion)\b/i,
  /\btransfer\b/i,
  /\bamend(?:ment)?\b/i,
  /\bterminate|termination\b/i,
];

const EXTERNAL_EXECUTION_PATTERNS = [
  /\bfile\b/i,
  /\bsubmit\b/i,
  /\bsend\b/i,
  /\bmail\b/i,
  /\bemail\b/i,
  /\btransmit\b/i,
  /\bexecute\b/i,
  /\bsign\b/i,
  /\brecord\b/i,
  /\bpublish\b/i,
  /\bserve\b/i,
  /\bopen (?:an )?account\b/i,
  /\bborrow\b/i,
  /\btransfer\b/i,
  /\bencumber\b/i,
  /\bdistribute\b/i,
];

function searchableRequestText(request: TaskRequest): string {
  let context = '';
  try {
    context = request.context ? JSON.stringify(request.context) : '';
  } catch {
    context = '';
  }
  return `${request.prompt}\n${context}`;
}

function readVerification(request: TaskRequest): TrustAuthorityVerification {
  const verification = request.context?.trustAuthority?.currentLawVerification;
  if (verification?.status === 'VERIFIED_CURRENT') {
    return {
      status: 'VERIFIED_CURRENT',
      jurisdiction: verification.jurisdiction,
      authorities: Array.isArray(verification.authorities) ? verification.authorities : [],
      verifiedAt: verification.verifiedAt,
      verifier: verification.verifier,
    };
  }
  if (verification?.status === 'CONFLICT_FOUND') {
    return {
      status: 'CONFLICT_FOUND',
      jurisdiction: verification.jurisdiction,
      authorities: Array.isArray(verification.authorities) ? verification.authorities : [],
      verifiedAt: verification.verifiedAt,
      verifier: verification.verifier,
    };
  }
  return { status: 'NOT_YET_VERIFIED' };
}

export function routeTrustAuthority(request: TaskRequest): TrustAuthorityRoute {
  const text = searchableRequestText(request);
  const isTrustRelated = TRUST_PATTERNS.some((pattern) => pattern.test(text));
  const legalEffectRequested = isTrustRelated && LEGAL_EFFECT_PATTERNS.some((pattern) => pattern.test(text));
  const externalExecutionRequested = isTrustRelated && EXTERNAL_EXECUTION_PATTERNS.some((pattern) => pattern.test(text));
  const currentLawVerification = readVerification(request);
  const principalApproval = request.context?.trustAuthority?.principalApproval === true;
  const blockingReasons: string[] = [];

  if (legalEffectRequested && currentLawVerification.status !== 'VERIFIED_CURRENT') {
    blockingReasons.push(
      'Current-law verification is required before a legal-effect conclusion or execution.',
    );
  }

  if (externalExecutionRequested && !principalApproval) {
    blockingReasons.push(
      'Principal/trustee approval is required before external trust execution.',
    );
  }

  return {
    isTrustRelated,
    routeId: 'HOWARD-TRUST-AUTHORITY',
    authorityOrder: [
      'trust-instrument-authority',
      'weisss-trustee-handbook',
      'current-law-verifier',
    ],
    legalEffectRequested,
    externalExecutionRequested,
    currentLawVerification,
    principalApproval,
    executionAllowed: blockingReasons.length === 0,
    blockingReasons,
  };
}

/**
 * Adds the mandatory authority route to planner context without weakening or
 * replacing caller-supplied context. This is how Hermes/SintraPrime planners
 * learn the required source order before any downstream reasoning.
 */
export function attachTrustAuthorityRoute(request: TaskRequest): TaskRequest {
  const route = routeTrustAuthority(request);
  if (!route.isTrustRelated) return request;

  return {
    ...request,
    context: {
      ...(request.context ?? {}),
      trustAuthorityRoute: route,
      mandatoryAuthorityInstructions: [
        'Consult trust-instrument-authority first for trust-specific governing language.',
        'Consult weisss-trustee-handbook second as secondary educational authority only.',
        'Obtain current-law verification before stating or executing any legal-effect conclusion.',
        'Do not execute an external trust action unless the required principal/trustee approval is documented.',
        'Record conflicts rather than silently reconciling lower-ranked sources with controlling authority.',
      ],
    },
  };
}

function isAuthorityResearchStep(step: PlanStep): boolean {
  const marker = `${step.description ?? ''} ${step.tool ?? ''} ${JSON.stringify(step.args ?? {})}`;
  return /trust-instrument-authority|weisss-trustee-handbook|current-law-verifier|verify current law|legal research/i.test(marker);
}

function isExternalStep(step: PlanStep): boolean {
  const marker = `${step.description ?? ''} ${step.tool ?? ''} ${JSON.stringify(step.args ?? {})}`;
  return EXTERNAL_EXECUTION_PATTERNS.some((pattern) => pattern.test(marker));
}

/**
 * Runtime gate evaluated immediately before every plan step.
 * Research/verification steps are always allowed so the system can satisfy the
 * gate. A step capable of creating legal effect remains blocked until current
 * law is verified; an external step also requires principal/trustee approval.
 */
export function evaluateTrustAuthorityStep(
  step: PlanStep,
  route: TrustAuthorityRoute,
): { allowed: boolean; reason?: string } {
  if (!route.isTrustRelated || isAuthorityResearchStep(step)) {
    return { allowed: true };
  }

  if (
    route.legalEffectRequested &&
    route.currentLawVerification.status !== 'VERIFIED_CURRENT'
  ) {
    return {
      allowed: false,
      reason: 'Trust authority gate: current law has not been verified.',
    };
  }

  if ((route.externalExecutionRequested || isExternalStep(step)) && !route.principalApproval) {
    return {
      allowed: false,
      reason: 'Trust authority gate: principal/trustee approval is not documented.',
    };
  }

  return { allowed: true };
}

File: channels/trust_authority_router.py
"""Trust authority dispatch metadata for Hermes/SintraPrime channel routing.

This module performs only deterministic request classification. It never claims
that current law has been verified. Downstream orchestration must satisfy the
current-law verifier and approval gates before a legal-effect conclusion or
external execution is allowed.
"""

from __future__ import annotations

import re
from typing import Any, Dict, Optional

_TRUST_PATTERNS = [
    re.compile(r"\bisiah\s+tarik\s+howard\s+trust\b", re.I),
    re.compile(r"\btrust instrument\b", re.I),
    re.compile(r"\bcertification of trust\b", re.I),
    re.compile(r"\bdeclaration of trust\b", re.I),
    re.compile(r"\btrustee\b", re.I),
    re.compile(r"\bco-?trustee\b", re.I),
    re.compile(r"\bbeneficiar(?:y|ies)\b", re.I),
    re.compile(r"\bsettlor\b", re.I),
    re.compile(r"\btrust corpus\b", re.I),
    re.compile(r"\bfiduciary dut(?:y|ies)\b", re.I),
    re.compile(r"\btrust administration\b", re.I),
    re.compile(r"\btrust banking\b", re.I),
]

_LEGAL_EFFECT_PATTERNS = [
    re.compile(r"\blegal effect\b", re.I),
    re.compile(r"\blegally binding\b", re.I),
    re.compile(r"\benforce(?:able|ment)?\b", re.I),
    re.compile(r"\bperfect(?:ion|ed)?\b", re.I),
    re.compile(r"\blien\b", re.I),
    re.compile(r"\bjurisdiction\b", re.I),
    re.compile(r"\btax(?:able|ation| filing| status)?\b", re.I),
    re.compile(r"\bcourt\b", re.I),
    re.compile(r"\bcreditor\b", re.I),
    re.compile(r"\bborrow(?:ing)?\b", re.I),
    re.compile(r"\bencumber\b", re.I),
    re.compile(r"\bdistribut(?:e|ion)\b", re.I),
    re.compile(r"\btransfer\b", re.I),
    re.compile(r"\bamend(?:ment)?\b", re.I),
]

_EXTERNAL_PATTERNS = [
    re.compile(r"\bfile\b", re.I),
    re.compile(r"\bsubmit\b", re.I),
    re.compile(r"\bsend\b", re.I),
    re.compile(r"\bmail\b", re.I),
    re.compile(r"\bemail\b", re.I),
    re.compile(r"\btransmit\b", re.I),
    re.compile(r"\bexecute\b", re.I),
    re.compile(r"\bsign\b", re.I),
    re.compile(r"\brecord\b", re.I),
    re.compile(r"\bpublish\b", re.I),
    re.compile(r"\bserve\b", re.I),
    re.compile(r"\bopen (?:an )?account\b", re.I),
]


def build_trust_authority_route(text: str) -> Optional[Dict[str, Any]]:
    """Return governed dispatch metadata for a trust-related message."""
    if not any(pattern.search(text) for pattern in _TRUST_PATTERNS):
        return None

    legal_effect = any(pattern.search(text) for pattern in _LEGAL_EFFECT_PATTERNS)
    external_execution = any(pattern.search(text) for pattern in _EXTERNAL_PATTERNS)

    return {
        "route_id": "HOWARD-TRUST-AUTHORITY",
        "authority_order": [
            "trust-instrument-authority",
            "weisss-trustee-handbook",
            "current-law-verifier",
        ],
        "legal_effect_requested": legal_effect,
        "external_execution_requested": external_execution,
        "current_law_status": "NOT_YET_VERIFIED",
        "principal_approval": False,
        "fail_closed": bool(legal_effect or external_execution),
    }
